hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestAbfsPaginatedDelete.java

```diff
@@ -0,0 +1,289 @@
+   * @throws Exception
+   */
+  @Test
+  public void testRecursiveDeleteWithInvalidCT() throws Exception {
+    testRecursiveDeleteWithInvalidCTInternal(true);
+    testRecursiveDeleteWithInvalidCTInternal(false);
+  }
+
+  private void testRecursiveDeleteWithPaginationInternal(boolean isEmptyDir,
+      boolean isPaginatedDeleteEnabled,
+      AbfsHttpConstants.ApiVersion xMsVersion) throws Exception {
+    final AzureBlobFileSystem fs = getUserFileSystem();
+    TracingContext testTracingContext = getTestTracingContext(fs, true);
+    Path testPath;
+    if (isEmptyDir) {
+      testPath = new Path("/emptyPath" + StringUtils.right(
+          UUID.randomUUID().toString(), 10));
+      fs.mkdirs(testPath);
+    } else {
+      testPath = createSmallDir();
+    }
+
+    // Set the paginated enabled value and xMsVersion at spiedClient level.
+    AbfsClient spiedClient = Mockito.spy(fs.getAbfsStore().getClient());
+    ITestAbfsClient.setAbfsClientField(spiedClient, "xMsVersion", xMsVersion);
+    Mockito.doReturn(isPaginatedDeleteEnabled).when(spiedClient).getIsPaginatedDeleteEnabled();
+
+    AbfsRestOperation op = spiedClient.deletePath(
+        testPath.toString(), true, null, testTracingContext);
+
+    // Getting the xMsVersion that was used to make the request
+    String xMsVersionUsed = getHeaderValue(op.getRequestHeaders(), X_MS_VERSION);
+    String urlUsed = op.getUrl().toString();
+
+    // Assert that appropriate xMsVersion and query param was used to make request
+    if (isPaginatedDeleteEnabled) {
+      Assertions.assertThat(urlUsed)
+          .describedAs("Url must have paginated = true as query param")
+          .contains(QUERY_PARAM_PAGINATED);
+      if (xMsVersion.compareTo(AbfsHttpConstants.ApiVersion.AUG_03_2023) < 0) {
+        Assertions.assertThat(xMsVersionUsed)
+            .describedAs("Request was made with wrong x-ms-version")
+            .isEqualTo(AbfsHttpConstants.ApiVersion.AUG_03_2023.toString());
+      } else if (xMsVersion.compareTo(AbfsHttpConstants.ApiVersion.AUG_03_2023) >= 0) {
+        Assertions.assertThat(xMsVersionUsed)
+            .describedAs("Request was made with wrong x-ms-version")
+            .isEqualTo(xMsVersion.toString());
+      }
+    } else {
+      Assertions.assertThat(urlUsed)
+          .describedAs("Url must not have paginated = true as query param")
+          .doesNotContain(QUERY_PARAM_PAGINATED);
+      Assertions.assertThat(xMsVersionUsed)
+          .describedAs("Request was made with wrong x-ms-version")
+          .isEqualTo(xMsVersion.toString());
+    }
+
+    // Assert that deletion was successful in every scenario.
+    AbfsRestOperationException e = intercept(AbfsRestOperationException.class, () ->
+        spiedClient.getPathStatus(testPath.toString(), false, testTracingContext, null));
+    Assertions.assertThat(e.getStatusCode())
+        .describedAs("Path should have been deleted").isEqualTo(HTTP_NOT_FOUND);
```

Comment:
* Factor this assertion out into a method here and below, something like `assertStatusCode(AbfsRestOperationException, int)`
* include e.toString() in message, or just rethrow the exception with its full stack trace.

Reply:
Makes sense.
Factored out to new method.
Added the e.toString() in message so that it will be a part of assertion error in case assertion fails.

Reply:
Taken